src/components/TextField/TextField.js

```diff
@@ -0,0 +1,317 @@
+    name: PropTypes.string,
+    onBlur: PropTypes.func,
+    onChange: PropTypes.func,
+    onEnterKey: PropTypes.func,
+    onFocus: PropTypes.func,
+    onKeyDown: PropTypes.func,
+    onKeyPress: PropTypes.func,
+    onKeyUp: PropTypes.func,
+    placeholder: PropTypes.string,
+    style: StyleObjectPropType(),
+    tooltipError: PropTypes.oneOfType([
+        PropTypes.element,
+        PropTypes.string,
+    ]),
+    tooltipHint: PropTypes.oneOfType([
+        PropTypes.element,
+        PropTypes.string,
+    ]),
+    tooltipRequired: PropTypes.oneOfType([
+        PropTypes.element,
+        PropTypes.string,
+    ]),
+    type: PropTypes.string,
+    value: PropTypes.string,
+    variant: ListPropType([
+        TextFieldVariant.DEFAULT,
+        TextFieldVariant.TITLE,
+    ]),
+};
+
+const defaultProps = {
+    className: null,
+    componentRef: null,
+    hasLabelAlways: false,
+    icon: null,
+    isClearable: false,
+    isDisabled: false,
+    isFullWidth: false,
+    isReadOnly: false,
+    isRequired: false,
+    isValid: null,
+    label: null,
+    name: null,
+    onBlur: null,
+    onChange: null,
+    onEnterKey: null,
+    onFocus: null,
+    onKeyDown: null,
+    onKeyPress: null,
+    onKeyUp: null,
+    placeholder: null,
+    style: null,
+    tooltipError: null,
+    tooltipHint: null,
+    tooltipRequired: 'required',
+    type: 'text',
+    value: '',
+    variant: TextFieldVariant.DEFAULT,
+};
+
+let lastInstanceId = 0;
+
+class TextField extends Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            hasFocus: false,
+            hasMouseOver: false,
+            showTooltip: false,
+            value: props.value,
+        };
+
+        lastInstanceId += 1;
+        this.uid = `textfield-${lastInstanceId}`;
+    }
+
+    handleMouseEnter = () => {
+        this.setState({ hasMouseOver: true });
+    }
+
+    handleMouseLeave = () => {
+        this.setState({ hasMouseOver: false });
+    }
+
+    handleInputRef = (ref) => {
+        const { componentRef } = this.props;
+        this.inputRef = ref;
+        if (componentRef) {
+            componentRef(ref);
+        }
+    }
+
+    handleInputBlur = (event) => {
+        const { onBlur } = this.props;
+        this.setState({
+            hasFocus: false,
+            showTooltip: false,
+        });
+        if (onBlur) {
+            onBlur(event);
+        }
+    }
+
+    handleInputChange = (event) => {
+        const { value } = this.inputRef;
+        const { onChange } = this.props;
+        if (typeof value !== 'undefined') {
+            this.setState({ value });
+        }
+        if (onChange) {
+            onChange(value, event);
+        }
+    }
+
+    handleInputFocus = (event) => {
+        const { onFocus } = this.props;
+        this.setState({
+            hasFocus: true,
+            showTooltip: true,
+        });
+        if (onFocus) {
+            onFocus(event);
+        }
```

Comment:
Same as my comments on TextArea, I think we should use `setState` callback here.